cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java

```diff
@@ -0,0 +1,126 @@
+ */
+public class ProgramRunStatusMonitorService extends AbstractIdleService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ProgramRunStatusMonitorService.class);
+
+  private final Store store;
+  private final ProgramRuntimeService runtimeService;
+  private final int txBatchSize;
+  private final CConfiguration cConf;
+  private ScheduledExecutorService programRunTerminateService;
+
+  @Inject
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService) {
+    this(cConf, store, runtimeService, cConf.getInt(Constants.AppFabric.PROGRAM_TERMINATOR_TX_BATCH_SIZE));
+  }
+
+  @VisibleForTesting
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService,
+                                 int txBatchSize) {
+    this.store = store;
+    this.runtimeService = runtimeService;
+    this.cConf = cConf;
+    this.txBatchSize = txBatchSize;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting ProgramRunStatusMonitorService");
+    programRunTerminateService = Executors
+      .newSingleThreadScheduledExecutor(new ThreadFactoryBuilder().setNameFormat("program run terminator").build());
+    long interval = cConf.getLong(Constants.AppFabric.PROGRAM_TERMINATOR_INTERVAL_SECONDS);
+    if (interval <= 0) {
+      LOG.warn("Invalid interval specified for the program terminator {}. Setting it to 3600 seconds.", interval);
+      interval = 3600L;
+    }
+    long initialDelay = cConf.getLong(Constants.AppFabric.PROGRAM_TERMINATOR_INITIAL_DELAY_SECONDS);
+    if (initialDelay <= 0) {
+      LOG.warn("Invalid initial delay specified for the local dataset deleter {}. Setting it to 300 seconds.",
+               initialDelay);
+      initialDelay = 300L;
+    }
+    programRunTerminateService.scheduleWithFixedDelay(this::terminateStoppingProgramsActiveBeyondTerminateTime,
+                                                      initialDelay, interval, TimeUnit.SECONDS);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Stopping ProgramRunStatusMonitorService");
+    programRunTerminateService.shutdown();
+    try {
+      if (!programRunTerminateService.awaitTermination(5, TimeUnit.SECONDS)) {
+        programRunTerminateService.shutdownNow();
+      }
+    } catch (InterruptedException ie) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  void terminateStoppingProgramsActiveBeyondTerminateTime() {
+    // fetch all runs that are in stopping state.
+    // Specifying the entire time range should not be worse in performance
+    // than specifying a more restrictive time range because time range is just used as a read-time filter
+    Map<ProgramRunId, RunRecordDetail> stoppingRuns = store.getRuns(ProgramRunStatus.STOPPING, 0L,
+                                                                    Long.MAX_VALUE, txBatchSize, null);
```

Comment:
The endtime shouldn't be max. It just need to be up to the current time minus some grace period.